### ISSUE TITLE
Gradle build sometimes gives spurious "unreferenced license file" warnings

### ISSUE THREAD

**Issue by mikemccand** [apache/lucene#13695]
### Description

On `main` today, with Java 21 on a recent Arch Linux, if I run `./gradlew clean` then `./gradlew check`, all is good.  It passes.

But if I run `./gradlew check` again, I sometimes (not always -- might take another `./gradlew check`) get these false warnings about unreferenced license files:

```
WARNING: there were unreferenced files under license folder:
  - /s1/l/9x/lucene/licenses/antlr4-runtime-LICENSE-BSD.txt
  - /s1/l/9x/lucene/licenses/antlr4-runtime-NOTICE.txt
  - /s1/l/9x/lucene/licenses/asm-LICENSE-BSD_LIKE.txt
  - /s1/l/9x/lucene/licenses/asm-NOTICE.txt
  - /s1/l/9x/lucene/licenses/asm-commons-LICENSE-BSD_LIKE.txt
  - /s1/l/9x/lucene/licenses/asm-commons-NOTICE.txt
  - /s1/l/9x/lucene/licenses/assertj-core-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/assertj-core-NOTICE.txt
  - /s1/l/9x/lucene/licenses/commons-codec-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/commons-codec-NOTICE.txt
  - /s1/l/9x/lucene/licenses/commons-compress-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/commons-compress-NOTICE.txt
  - /s1/l/9x/lucene/licenses/commons-logging-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/commons-logging-NOTICE.txt
  - /s1/l/9x/lucene/licenses/hamcrest-LICENSE-BSD.txt
  - /s1/l/9x/lucene/licenses/hamcrest-NOTICE.txt
  - /s1/l/9x/lucene/licenses/httpclient-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/httpclient-NOTICE.txt
  - /s1/l/9x/lucene/licenses/httpcore-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/httpcore-NOTICE.txt
  - /s1/l/9x/lucene/licenses/icu4j-LICENSE-BSD_LIKE.txt
  - /s1/l/9x/lucene/licenses/icu4j-NOTICE.txt
  - /s1/l/9x/lucene/licenses/javax.servlet-LICENSE-CDDL.txt
  - /s1/l/9x/lucene/licenses/javax.servlet-NOTICE.txt
  - /s1/l/9x/lucene/licenses/jetty-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/jetty-NOTICE.txt
  - /s1/l/9x/lucene/licenses/jts-core-LICENSE-BSD_LIKE.txt
  - /s1/l/9x/lucene/licenses/jts-core-NOTICE.txt
  - /s1/l/9x/lucene/licenses/junit-LICENSE-CPL.txt
  - /s1/l/9x/lucene/licenses/junit-NOTICE.txt
  - /s1/l/9x/lucene/licenses/morfologik-fsa-LICENSE-BSD.txt
  - /s1/l/9x/lucene/licenses/morfologik-fsa-NOTICE.txt
  - /s1/l/9x/lucene/licenses/morfologik-polish-LICENSE-BSD.txt
  - /s1/l/9x/lucene/licenses/morfologik-polish-NOTICE.txt
  - /s1/l/9x/lucene/licenses/morfologik-stemming-LICENSE-BSD.txt
  - /s1/l/9x/lucene/licenses/morfologik-stemming-NOTICE.txt
  - /s1/l/9x/lucene/licenses/morfologik-ukrainian-search-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/morfologik-ukrainian-search-NOTICE.txt
  - /s1/l/9x/lucene/licenses/nekohtml-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/nekohtml-NOTICE.txt
  - /s1/l/9x/lucene/licenses/opennlp-tools-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/opennlp-tools-NOTICE.txt
  - /s1/l/9x/lucene/licenses/procfork-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/procfork-NOTICE.txt
  - /s1/l/9x/lucene/licenses/randomizedtesting-runner-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/randomizedtesting-runner-NOTICE.txt
  - /s1/l/9x/lucene/licenses/s2-geometry-library-java-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/s2-geometry-library-java-NOTICE.txt
  - /s1/l/9x/lucene/licenses/spatial4j-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/spatial4j-NOTICE.txt
  - /s1/l/9x/lucene/licenses/xercesImpl-LICENSE-ASL.txt
  - /s1/l/9x/lucene/licenses/xercesImpl-NOTICE.txt
```

It's sometimes a smaller set of (still false positive) warnings, which is also odd.

For the longest time I've been assuming I'm doing something wrong :)  But then I noticed @msokolov also sees this sometimes, so it's not just me.  Does anyone else see these?

[I also raised [this dev list discussion](https://lists.apache.org/thread/xgqb00w597z087fsf9sj2s7jq27y2llc) about it].

**Comment by uschindler** [comment]
I alos noticed this several times. I also reported an issue to Dawid, but the cause for this is very strage. It depends on the order of execution.

I used to know the cause of this bug, but I was unable to fix it.

**Comment by msokolov** [comment]
Just curious - what is the purpose of the check? I guess it sounds self-explanatory: if there is a license file thatisn't referenced (where?) we would presumably want to clean it up, but I wonder how often this is an issue relative to the false positives? Should we disable until we can fix?

**Comment by uschindler** [comment]
> Just curious - what is the purpose of the check? I guess it sounds self-explanatory: if there is a license file thatisn't referenced (where?) we would presumably want to clean it up, but I wonder how often this is an issue relative to the false positives? Should we disable until we can fix?

No the check is very valid. It should warn you if you have removed or added a dependency but its corresponding license file is still there.

I think I remember the issue: The cause is in the following file: https://github.com/apache/lucene/blob/d55b92bae01be46656c4d373d8cb1d081b280469/gradle/validation/jar-checks.gradle

The problem is that the task reporting this failure sometimes has the "allReferenced" empty, so it reports all existing as dangling.

I think it has to do if the `ext.jarInfos` was not yet filled with information. On build setup this is initialized to be empty: https://github.com/apache/lucene/blob/d55b92bae01be46656c4d373d8cb1d081b280469/gradle/validation/jar-checks.gradle#L61

Later it expects the task `collectJarInfos` to populate this. But for some reason as this task does not seem to have inputs/outputs configured it sometimes does not run. So the check tasks repots all license files as missing, because this suddenly substracts the empty list and everything is left over:

https://github.com/apache/lucene/blob/d55b92bae01be46656c4d373d8cb1d081b280469/gradle/validation/jar-checks.gradle#L350

Basically, it always fails on a second run of gradle when the gradle daemon is still active and has cached the task's output somehow, but actually the task output is not restored into the ext property.

**Comment by uschindler** [comment]
I think to fix this, the `collectJarInfos` must write that information to a monitored file in the build outputs.

All pother tasks using that information must read the output from the file and not work with a global property which is not part of the monitored set of build outputs.

**Comment by uschindler** [comment]
Let's ask @dweiss, I gave up on this, as my Gradle-Foo is too limited,

**Comment by uschindler** [comment]
Maybe it can be solved by making the "ext" property a build output and let all tasks depend on it.

**Comment by uschindler** [comment]
P.S.: I did not open an issue because the last time I was confronted with it (about 1 year ago), I was at the mentag stage of "damn fcking gradle - I need more beers", but my fridge was empty of beers (like the ext property) and I went to bed and ignored the issue on following day.

**Comment by risdenk** [comment]
Solr has the same issue and it was reported here https://issues.apache.org/jira/browse/SOLR-17142 - there was a mailing list post where I found the culprit but I haven't had a chance to figure out the gradle isms.

**Comment by uschindler** [comment]
OK, I debugged a bit:

```
>gradlew :lucene:checkDanglingLicenseFiles
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :errorProneSkipped
WARNING: errorprone disabled (skipped on builds not running inside CI environments, pass -Pvalidation.errorprone=true to enable)

> Task :lucene:core:compileJava
Hinweis: Einige Eingabedateien verwenden oder ├╝berschreiben eine veraltete API.
Hinweis: Wiederholen Sie die Kompilierung mit -Xlint:deprecation, um Details zu erhalten.

> Task :lucene:facet:compileJava
Hinweis: Einige Eingabedateien verwenden oder ├╝berschreiben eine veraltete API.
Hinweis: Wiederholen Sie die Kompilierung mit -Xlint:deprecation, um Details zu erhalten.

> Task :lucene:queries:compileJava
Hinweis: C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\queries\src\java\org\apache\lucene\queries\payloads\PayloadMatcherFactory.java verwendet oder ├╝berschreibt eine veraltete API.
Hinweis: Wiederholen Sie die Kompilierung mit -Xlint:deprecation, um Details zu erhalten.

> Task :lucene:sandbox:compileJava
Hinweis: Einige Eingabedateien verwenden oder ├╝berschreiben eine veraltete API.
Hinweis: Wiederholen Sie die Kompilierung mit -Xlint:deprecation, um Details zu erhalten.

> Task :lucene:analysis:common:compileJava
Hinweis: Einige Eingabedateien verwenden oder ├╝berschreiben eine veraltete API.
Hinweis: Wiederholen Sie die Kompilierung mit -Xlint:deprecation, um Details zu erhalten.

> Task :lucene:expressions:compileJava
Hinweis: C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\expressions\src\java\org\apache\lucene\expressions\js\JavascriptCompiler.java verwendet oder ├╝berschreibt eine veraltete API.
Hinweis: Wiederholen Sie die Kompilierung mit -Xlint:deprecation, um Details zu erhalten.

> Task :lucene:test-framework:compileJava
Hinweis: Einige Eingabedateien verwenden oder ├╝berschreiben eine veraltete API.
Hinweis: Wiederholen Sie die Kompilierung mit -Xlint:deprecation, um Details zu erhalten.

> Task :lucene:grouping:compileJava
Hinweis: C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\grouping\src\java\org\apache\lucene\search\grouping\GroupingSearch.java verwendet oder ├╝berschreibt eine veraltete API.
Hinweis: Wiederholen Sie die Kompilierung mit -Xlint:deprecation, um Details zu erhalten.

BUILD SUCCESSFUL in 1m 4s
208 actionable tasks: 185 executed, 23 up-to-date
<-------------> 0% WAITING
> IDLE
> IDLE
> IDLE
> IDLE


>gradlew :lucene:checkDanglingLicenseFiles

> Task :errorProneSkipped
WARNING: errorprone disabled (skipped on builds not running inside CI environments, pass -Pvalidation.errorprone=true to enable)
WARNING: there were unreferenced files under license folder:
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\antlr4-runtime-LICENSE-BSD.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\antlr4-runtime-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\asm-LICENSE-BSD_LIKE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\asm-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\asm-commons-LICENSE-BSD_LIKE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\asm-commons-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\assertj-core-LICENSE-ASL.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\assertj-core-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\commons-codec-LICENSE-ASL.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\commons-codec-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\commons-compress-LICENSE-ASL.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\commons-compress-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\hamcrest-LICENSE-BSD.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\hamcrest-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\icu4j-LICENSE-BSD_LIKE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\icu4j-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\jts-core-LICENSE-BSD_LIKE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\jts-core-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\junit-LICENSE-CPL.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\junit-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\morfologik-fsa-LICENSE-BSD.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\morfologik-fsa-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\morfologik-polish-LICENSE-BSD.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\morfologik-polish-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\morfologik-stemming-LICENSE-BSD.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\morfologik-stemming-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\morfologik-ukrainian-search-LICENSE-ASL.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\morfologik-ukrainian-search-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\nekohtml-LICENSE-ASL.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\nekohtml-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\opennlp-tools-LICENSE-ASL.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\opennlp-tools-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\procfork-LICENSE-ASL.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\procfork-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\randomizedtesting-runner-LICENSE-ASL.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\randomizedtesting-runner-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\s2-geometry-library-java-LICENSE-ASL.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\s2-geometry-library-java-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\slf4j-api-LICENSE-MIT.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\spatial4j-LICENSE-ASL.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\spatial4j-NOTICE.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\xercesImpl-LICENSE-ASL.txt
  - C:\Users\Uwe Schindler\Projects\lucene\lucene\lucene\licenses\xercesImpl-NOTICE.txt

BUILD SUCCESSFUL in 3s
208 actionable tasks: 83 executed, 125 up-to-date
```

I figured out that stopping gradle daemons does not work. The main issue is a different problem:
When everything is already compiled, the `collectJarInfos` task is not executed (for no subproject) where the outputs exist.

The problem is the dynamic collection of all tasks that the `checkDanglingLicenseFiles` task only depends on this dynamically generated 'validationTasks`, which may not execute the `jarInfoTasks`.

When you clean, it works again.

**Comment by uschindler** [comment]
OK, I found the bug. The comment explains it:

https://github.com/apache/lucene/blob/d55b92bae01be46656c4d373d8cb1d081b280469/gradle/validation/jar-checks.gradle#L138-L140

Basically when the real check tasks are not executed at all (due to everything uptodated), `referencedFiles` stays empty and then it complains about the files as those are not removed form the files found in the directory.

We have to make sure that all of the tasks which change the referencedFiles property actually run.

**Comment by uschindler** [comment]
If I comment out those 2 lines, it works:
https://github.com/apache/lucene/blob/d55b92bae01be46656c4d373d8cb1d081b280469/gradle/validation/jar-checks.gradle#L201-L202

So we have to decide: Unbreak this and don't allow incremental builds for the license check or fix it fully?

This was broken by, so I'd like to remove that optimization: https://github.com/apache/lucene/pull/12150

**Comment by mikemccand** [comment]
Wow, thank you for digging so deep @uschindler!  Maybe your fridge has beers again!

**Comment by mikemccand** [comment]
+1 to revert that part of the optimization (no incremental building of license checking) for now?

How slow is the full check?

**Comment by uschindler** [comment]
It's fast. I see no delay.

**Comment by uschindler** [comment]
Here is the PR: #13696

I added a comment to the top of this file and a TODO to explain that all tasks in this file must always run and therefore cannot define any outputs.

I will merge this soon. I think Solr needs to do the same to fix this issue. @risdenk you can mention this PR in https://issues.apache.org/jira/browse/SOLR-17142

**Comment by uschindler** [comment]
@mikemccand: Beer is ready and after having a shower in my German bathroom the hot day can end!

**Comment by dweiss** [comment]
> after having a shower in my German bathroom

I'm sure all of us can't stop picturing this German bathroom, Uwe! :)

**Comment by uschindler** [comment]
> > after having a shower in my German bathroom
> 
> I'm sure all of us can't stop picturing this German bathroom, Uwe! :)

https://x.com/UweSays/status/91539616850644992?t=_DLF7fly9WrgH6wQ15M13A&s=19

**Comment by mikemccand** [comment]
> > > after having a shower in my German bathroom
> > 
> > 
> > I'm sure all of us can't stop picturing this German bathroom, Uwe! :)
> 
> https://x.com/UweSays/status/91539616850644992?t=_DLF7fly9WrgH6wQ15M13A&s=19

I know which US shower @uschindler was referring to there!!

**Comment by mikemccand** [comment]
Thank you @uschindler!
